src/Netscript/APIWrapper.ts

```diff
@@ -0,0 +1,157 @@
+import { getRamCost } from "./RamCostGenerator";
+import type { IPort } from "../NetscriptPort";
+import type { BaseServer } from "../Server/BaseServer";
+import type { WorkerScript } from "./WorkerScript";
+import { makeRuntimeRejectMsg } from "../NetscriptEvaluator";
+import { Player } from "../Player";
+
+type ExternalFunction = (...args: any[]) => any;
+type ExternalAPI = {
+  [string: string]: ExternalAPI | ExternalFunction;
+};
+
+type InternalFunction<F extends (...args: unknown[]) => unknown> = (
+  ctx: NetscriptContext,
+) => (...args: unknown[]) => ReturnType<F>;
+export type InternalAPI<API> = {
+  [Property in keyof API]: API[Property] extends ExternalFunction
+    ? InternalFunction<API[Property]>
+    : API[Property] extends ExternalAPI
+    ? InternalAPI<API[Property]>
+    : never;
+};
+
+type WrappedNetscriptFunction = (...args: unknown[]) => unknown;
+type WrappedNetscriptAPI = {
+  [string: string]: WrappedNetscriptAPI | WrappedNetscriptFunction;
+};
+
+export type NetscriptContext = {
```

Comment:
nitpick, some of these field should be marked as readonly

Reply:
adding a commit in a moment